File: Python/bit-operation/number_of_valid_words_for_each_puzzle.py
import collections
from typing import List

def findNumOfValidWords(words: List[str], puzzles: List[str]) -> List[int]:
    frequency = collections.Counter()
    for word in words:
        mask = 0
        for ch in word:
            mask |= (1 << (ord(ch) - ord("a")))
        if str(bin(mask)).count("1") <= 7:
            frequency[mask] += 1
    
    ans = list()
    for puzzle in puzzles:
        total = 0
        mask = 0
        for i in range(1, 7):
            mask |= (1 << (ord(puzzle[i]) - ord("a")))
        
        subset = mask
        while subset:
            s = subset | (1 << (ord(puzzle[0]) - ord("a")))
            if s in frequency:
                total += frequency[s]
            subset = (subset - 1) & mask
        
        if (1 << (ord(puzzle[0]) - ord("a"))) in frequency:
            total += frequency[1 << (ord(puzzle[0]) - ord("a"))]
        
        ans.append(total)
    return ans

File: core/tmp/Java/src/com/dbc/algorithm/BitOperation/NumberOfValidWordsForEachPuzzle.java
package com.dbc.algorithm.BitOperation;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class NumberOfValidWordsForEachPuzzle {
    public static List<Integer> findNumOfValidWords(String[] words, String[] puzzles) {
        Map<Integer, Integer> map = new HashMap<>();

        for (String word : words) {
            int mask = 0;
            for (int i = 0; i < word.length(); i++) {
                mask |= (1 << (word.charAt(i) - 'a'));
            }
            map.put(mask, map.getOrDefault(mask, 0) + 1);
        }

        List<Integer> result = new ArrayList<>();
        for (String puzzle : puzzles) {
            int mask = 0, total = 0;
            for (int i = 1; i < puzzle.length(); i++) {
                mask |= (1 << (puzzle.charAt(i) - 'a'));
            }

            int subset = mask;
            while (subset != 0) {
                int s = subset | (1 << (puzzle.charAt(0) - 'a'));
                if (map.containsKey(s)) {
                    total += map.get(s);
                }
                subset = (subset - 1) & mask;
            }
            if (map.containsKey((1 << (puzzle.charAt(0) - 'a')))) {
                total += map.get((1 << (puzzle.charAt(0) - 'a')));
            }
            result.add(total);
        }
        return result;
    }
}
